[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc

mikaylagawarecki opened the pull request:
Adds a path with the strategy mentioned [here](https://github.com/pytorch/pytorch/pull/95346#issuecomment-1441283506) 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #95657
* #95486

mikaylagawarecki committed:
[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc

[ghstack-poisoned]

drisspg commented on aten/src/ATen/native/transformers/cuda/sdp_utils.h:
```diff
@@ -185,6 +186,7 @@ inline bool check_for_seq_len_1_nested_tensor(sdp_params params, bool debug) {
 
 inline bool check_for_nested_inputs(sdp_params params, bool debug){
   if (params.query.is_nested() || params.key.is_nested() || params.value.is_nested()) {
+    // FIXME: this debug message should be removed now
```
I will move this into the other PR

drisspg commented on aten/src/ATen/native/transformers/cuda/sdp_utils.h:
```diff
@@ -56,7 +57,7 @@ inline std::array<SDPBackend, num_backends> priority_order(sdp_params params) {
   // MemEff parallelizes across "batch_size * num_heads * num_queries" and can
   // be more efficient. batch_size, q_len, num_heads, k = inp.query.shape
-  if (params.query.is_nested()) {
+  if (params.key.is_nested()) {
```
maybe we should q,k,v looking for nested

drisspg commented on aten/src/ATen/native/transformers/cuda/sdp_utils.h:
```diff
@@ -246,21 +248,77 @@ inline bool check_tensor_shapes(sdp_params params, bool debug) {
     if (debug) {
       TORCH_WARN(
-        "Both fused kernels requires query, key and value to have the same batch_size and num_heads. Query.sizes(): ",
+        "For dense inputs, both fused kernels require query, key and value to have the same batch_size and num_heads. ",
```
Might be worth saying, try unsqueeze and expand_to but I think that if they are savy enough to turn off the math fallback they will probably know that.

drisspg commented on aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:
```diff
@@ -319,78 +319,234 @@ bool is_safe_to_get_storage_as_tensor(const NestedTensorImpl* tensor) {
+  if (k_batch_size_needs_broadcast) {
+    if (v_batch_size_needs_broadcast) {
+      // Should not be the case that both k and v need broadcasting of batch_size
+      // i.e. output_batch_size == 1 and k/v_batch_size are both 1
```
!= 1 

Also is this checked in sdp_utils?

mikaylagawarecki replied:
Will add a check, thanks for catching

mikaylagawarecki replied:
edit: actually @drisspg, do you think this case output_batch_size == 1 and k/vbatch_size == 1 might potentially sense? rethinking it now I I might have missed it here cuz then it would just be

k, v `[1, num_head, x, embed_qk], [1, num_head, x, embed_v] --> [B, num_head, x, embed_qk], [B, num_head, x, embed_v]`

The case I actually missed was if 1 of k/v batch_size needs broadcast and not the other then the other must have a consistent seq_len dim, e.g. 

k, v `[1, num_head, x, embed_qk], [B, num_head, x, embed_v] --> [B, num_head, x, embed_qk], [B, num_head, x, embed_v]`

does that sound right to you?

drisspg replied:
To be honest this is the most logic dense part of the code and little hard to follow all the paths. But the broadcasting examples above I agree with. Like you said in the second example when k and v are nested we need to be more careful on the seq_len broadcasting

drisspg commented on test/test_transformers.py:
```diff
@@ -1963,6 +1952,36 @@ def test_fused_kernels_seq_len_0_inputs(self):
+        key = rand_nested_tensor(k_shape)
+        value = rand_nested_tensor(v_shape)
+
+        query_expanded = torch.nested.nested_tensor([query.squeeze(0) for _ in range(batch)]).transpose(1, 2)
```
might be nice to add a  comment of the nested shapes
ie.e # query_expended = (num_heads, 1, head_dim)

mikaylagawarecki committed:
Update on "[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc"

Adds a path with the strategy mentioned [here](https://github.com/pytorch/pytorch/pull/95346#issuecomment-1441283506) 




[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/nested/NestedTensorMatmul.cpp:
```diff
@@ -306,7 +306,7 @@ Tensor matmul_nested(const Tensor& self, const Tensor& mat2) {
       mat2_dim == 4 && mat2.is_contiguous() &&
       !(GradMode::is_enabled() && (self.requires_grad() || mat2.requires_grad()))) {
-    auto n_heads = self_sizes.select(0, 1).select(0, 0).item<int64_t>();
+    auto n_heads = self_sizes.select(0, 0).select(0, 0).item<int64_t>();
```
Is this fixing a bug? If so let's split it out into a separate PR and update the corresponding matmul tests to cover it.

drisspg replied:
Also can't we use opt_dim to do this?

mikaylagawarecki committed:
Update on "[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc"

Adds a path with the strategy mentioned [here](https://github.com/pytorch/pytorch/pull/95346#issuecomment-1441283506) 




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc"

Adds a path with the strategy mentioned [here](https://github.com/pytorch/pytorch/pull/95346#issuecomment-1441283506) 




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc"

Adds a path with the strategy mentioned [here](https://github.com/pytorch/pytorch/pull/95346#issuecomment-1441283506) 




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc"

Adds a path with the strategy mentioned [here](https://github.com/pytorch/pytorch/pull/95346#issuecomment-1441283506) 




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc"

Adds a path with the strategy mentioned [here](https://github.com/pytorch/pytorch/pull/95346#issuecomment-1441283506) 




[ghstack-poisoned]

drisspg commented on aten/src/ATen/native/transformers/cuda/sdp_utils.h:
```diff
@@ -246,27 +314,77 @@ inline bool check_tensor_shapes(sdp_params params, bool debug) {
+  auto v_batch_size = params.value.size(0);
+
+  // Flip because check_for_nested_inputs returns false if there exists a nested input
+  bool has_nested_input = !check_for_nested_inputs(params);
```
Nit: since we are just using this as a helper now and not a filter_function we could probs switch this logic, but no worries

drisspg commented on aten/src/ATen/native/transformers/cuda/sdp_utils.h:
```diff
@@ -246,27 +314,77 @@ inline bool check_tensor_shapes(sdp_params params, bool debug) {
+      // must have a consistent seq_len dim
+      if (broadcastable_batch_size) {
+        if (k_batch_size == 1 && v_batch_size != 1) {
+          const auto v_nt_tensor_impl = at::native::get_nested_tensor_impl(params.value);
```
maybe pull this little chunk out and apply on v and then k if needed

drisspg commented on test/test_nestedtensor.py:
```diff
@@ -1861,27 +1862,28 @@ def rand_mask(size):
+                    query, key, value, attn_mask=attn_mask, dropout_p=dropout_p)
+
+            # Error case: is_causal=True.
+            with self.assertRaisesRegex(RuntimeError, "not supported when is_causal=True"):
```
I want to check this, cause I don't know if this fails with the fused kernels

mikaylagawarecki replied:
I think I shouldn't have indented this code block with a `sdp_kernel(enable_flash=False, enable_math=True, enable_mem_efficient=False)`, both num_head and seq_len are ragged here so it should always dispatch to math, but I should still be checking that it doesn't dispatch to the wrong kernel, my bad

drisspg reviewed (approved):
Awesome!

mikaylagawarecki committed:
Update on "[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc"

Adds a path with the strategy mentioned [here](https://github.com/pytorch/pytorch/pull/95346#issuecomment-1441283506) 




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[sdpa] Add broadcasting for batch and num_heads dimensions to fused kernel nested preproc"

Adds a path with the strategy mentioned [here](https://github.com/pytorch/pytorch/pull/95346#issuecomment-1441283506) 




[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge
